### ISSUE TITLE
Implement Switching

### ISSUE THREAD

**Issue by ash47** [ash47/LegendsOfDota#274]

**Comment by darklordabc** [comment]
I closed the other one? I didn't intend to.

**Comment by darklordabc** [comment]
Okay, I think I have friend who can do this feature.

Heres the current plan.
Add a small GUI component on screen, that will offer a list of players to switch with. This is an example of how it could work (Obviously not in that position, and not with those two buttons):
![options2](https://cloud.githubusercontent.com/assets/16277198/15136071/eadcca7c-16c0-11e6-83c4-dfdc72687880.gif)

After you choose a player, it will bring up a vote for the other players:
![image](https://cloud.githubusercontent.com/assets/16277198/15136083/06e504b4-16c1-11e6-9cd4-e4ecbdf88501.png)

And if a certain % of players agree (maybe all minus 1 or 2), you will switch and spawn at the other fountain. 

Thoughts?

**Comment by ash47** [comment]
My current plan would be that once teams become unbalanced, a notification
will come up that allows players to nominate themselves to fix the balance
issues, once you nominate yourself, a vote will be created to swap you to
the other team, you need 50% + 1 vote in order to pass the vote, if it
passes, the player will be swapped.

**Comment by darklordabc** [comment]
What if we combine the best of both solutions. I like your idea of the auto notification, and the "nominate yourself" button, however, what if players are afk but still in the game, or if the teams are unfair but still not uneven enough to trigger the notification? Like if its 4 vs 5, and the 5 team is dominating, and the 4 team has basically 2 retards on it. In these situations you may want to initiate a switch manually.

So how about, we have the notification system and the nominate yourself option, but also have a switch option that you can manually call a switch? I'll get my friend to do.

**Comment by Jexah** [comment]
Introducing team swapping with regards to bad players is a bad idea. The only reason for team swapping is to balance the numbers, not the players' skill levels. To balance for afkers, we'd first need to determine that they are afk, then kick them and give them an abandon. 

Additionally, Ash, I'd prefer if the player vote requirement for swapping would be 50% + 1 for each team, so that the team with less players can protect themselves in regards to gold and exp scaling.

**Comment by darklordabc** [comment]
Player swapping could require unanimous approval.

**Comment by darklordabc** [comment]
Okay I've done some thinking about this, and I'd like some feedback on what I've thought up so far.

The switch button could be a subtle inclusion to the top of the screen, where the other team management buttons are. Its subtle, but not too subtle that you wouldn't notice it. 

![image](https://cloud.githubusercontent.com/assets/16277198/15149554/49161822-170d-11e6-8657-ae5b879d1201.png)

When the teams become unbalanced, such as after a player abandons, and it is now 5 vs 3, or 4 vs 1.
The switch button could flash, and a little text box could appear informing players that now is a good time to nominate to switch. 

![image01](https://cloud.githubusercontent.com/assets/16277198/15149598/8a0caf08-170d-11e6-8da0-0646587c4007.gif)

When players click the flashing switch button, it will auto nominate them to switch to one of the random disconnected players (or better yet the lowest level disconnected player). However, if the player clicks the switch button during its resting state, and not when its flashing, a switch menu should appear, allowing for manual switching. 

![image](https://cloud.githubusercontent.com/assets/16277198/15149679/dd19e53a-170d-11e6-8642-b1f102aed093.png)

The switch popup box should stay visible for as long as the mouse cursor stays in this bounding box.
 
![image](https://cloud.githubusercontent.com/assets/16277198/15149741/0e368894-170e-11e6-9840-a5bc5e24a684.png)

Once a player is nominated to switch, either by clicking the switch while its flashing, or by manually picking to switch, the game will pause (using the code "PauseGame(true)"), and a vote prompt will come up in the top left hand corner under the switch button, telling users who has nominated to switch with who, and the option to vote for yes or no. There should also be a timer (like 10 seconds). 

![image](https://cloud.githubusercontent.com/assets/16277198/15149951/dce8768e-170e-11e6-962a-c353d9d25146.png)

The formula for deciding if the vote passes or not is designed so that it gets easier to pass, the more unbalanced the teams are. For a vote to pass it must have:
"Yes" votes equal to the number of players in game minus 2. 
The allowed of "No" votes tolerated before a vote fails, is the player difference of teams minus 1. 
So if there are 8 players in the game, and the teams are 5 vs 3, a vote would require 6 "Yes" (8 - 2) and no more than 1 "No" (2 - 1). The scales should be against switching by default, as you don't want people to get burnt by it, it should only be used to save games. 

For player vs player swaps, the votes must have the number of players in the game, minus 1, and zero "No". Essentially, everybody has to be onboard. 

More examples:
Player for Disconnected Player, From stronger team to losing team.
5 vs 4 =  7 Approve, 0 No
5 vs 3 = 6 Approve, <= 1 No
5 vs 2 = 5 Approve, <= 2 No
5 vs 1 = 4 Approve, <= 3 No

4 vs 3 = 5 Approve, 0 No
4 vs 2 = 4 Approve, <= 1 No
4 vs 1 = 3 Approve, <= 2 No

3 vs 2 = 3 Approve, 0 No
3 vs 1 = 2 Approve, <= 1 No

Player for Player
5 vs 5 = 9 Approve, 0 No
5 vs 4 = 8 Approve, 0 No
5 vs 3 = 7 Approve, 0 No
5 vs 2 = 6 Approve, 0 No
5 vs 1 = 5 Approve, 0 No

**Comment by darklordabc** [comment]
Alternatively we could cut out manual swapping and have the button greyed out when no imbalance present.

**Comment by darklordabc** [comment]
@Ash47 can you please confirm that this is good to go ahead with? I've got my friend programmer lined up to do this, and this feature is needed quite urgently.

**Comment by ash47** [comment]
@jexah and I don't really like the idea of people switching teams, and on
top of that, you can't actually go above the team limits, that is it say if
it's 5vs5, you can't trade someone over so it becomes 6vs4 even if you will
even it out straight after.

The only reason the swapping works with people who have abandoned is
because it allows 6vs4 because really it's 5vs4 being swapped to 4vs5.

The swapping code has been implemented in the ingame.lua file.

Please also note that all new UI needs to be built using object oriented
code, the selection ui will eventually be updated to this standard now that
I know how to make the components talk to each other.

**Comment by darklordabc** [comment]
I wont do player switching, just swapping with disconnected players when there is at least a 2 player difference. As for the new UI, can you be more specific? Is there some example interface we can look at? I'm sure my friend can program it in any language.

**Comment by ash47** [comment]
It's HTML, CSS, and JavaScript

**Comment by darklordabc** [comment]
Is any of the current UI build in this interface? But I'll get started none the less, on adding that button to switch.

**Comment by Jazzer008** [comment]
Progress so far. 
![](http://i.imgur.com/8ZsqcEZ.png)

**Comment by darklordabc** [comment]
Looks great so far

**Comment by Jazzer008** [comment]
![](http://i.imgur.com/Z44metZ.jpg)
![](http://i.imgur.com/Oqw4Gwl.jpg)
![](http://i.imgur.com/nryM8Cv.jpg)

**Comment by Jazzer008** [comment]
@ash47 Currently, your team switching method involves 'custom' team assignments. I take these into account when displaying the team switch information but the top scorebar icons and actual tab scoreboard still display incorrect team information. I'm think those js files need to be updated with the Custom team assignemnts and not the playerInfo.player_team_id.

Anyhoo, apart from the voting to switch. Everything else is working as assumed so far.

**Comment by darklordabc** [comment]
Looking good so far. Hopefully Ash can get around to looking into this as its still sorely needed.

**Comment by ash47** [comment]
Oh? Is this a pull request? Assumed it was an issue :O

**Comment by Jazzer008** [comment]
@ash47 No, not yet. I was just hoping for some input on my questions, ponderings?

**Comment by ash47** [comment]
@jexah

**Comment by Myrl** [comment]
The implementation itself seems complete but there are still some things that are debatable.
